src/foam/core/stdlib.js

```diff
@@ -746,7 +774,17 @@ foam.LIB({
     methods: [
       function clone(o)      { return typeOf(o).clone(o); },
       function equals(a, b)  { return typeOf(a).equals(a, b); },
-      function compare(a, b) { return typeOf(a).compare(a, b); },
+      function compare(a, b) {
+        // To ensure that symmetry is present when comparing,
+        // we will always use the comparator of higher precedence.
+        var aOrdinal = typeOf(a).ordinal;
+        var bOrdinal = typeOf(b).ordinal;
+        foam.assert(aOrdinal !== undefined && bOrdinal !== undefined,
+            'comparator cannot operate on unknown types');
+
+        return (aOrdinal > bOrdinal) ? typeOf(a).compare(a, b) :
```

Comment:
typeOf() is expensive, which is why in the code I proposed I stored type type in a variable rather than the ordinal, so that I wouldn't have to look it up again here.